src/main/resources/category/java/common.xml

```diff
@@ -1378,4 +1378,403 @@ class Foo {
+                ]]></value>
+            </property>
+        </properties>
+        <example>
+            <![CDATA[
+class Bad {
+    String field1;
+
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Bad that = (Bad) o;
+        return Objects.equals(field1, that.field1);
+    }
+    public int hashCode() {
+        return super.hashCode(); // bad
+    }
+}
+
+class Good {
+    String field1;
+
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Bad1 that = (Bad1) o;
+        return Objects.equals(field1, that.field1);
+    }
+    public int hashCode() {
+        return Objects.hash(field1); // good
+    }
+}
+            ]]>
+        </example>
+    </rule>
+
+    <rule name="AvoidRecreatingSecurityProviders"
+          language="java"
+          message="Avoid re-creating security providers, this is expensive."
+          class="net.sourceforge.pmd.lang.rule.xpath.XPathRule"
+
+          externalInfoUrl="${doc_root}/JavaCodePerformance.md#iuosf01">
+        <description>Problem: Creating a security provider is expensive because of loading of algorithms and other classes. Additionally, it uses synchronized which leads to lock contention when used with multiple threads.
+            Solution: This only needs to happen once in the JVM lifetime, because once loaded, the provider is typically available from the Security class. Create the security provider only once: only in case it is nog available from the Security class, yet.
+        </description>
+        <priority>2</priority>
+        <properties>
+            <property name="tags" value="jpinpoint-rule,performance" type="String" description="classification"/>
+            <property name="xpath">
+                <value><![CDATA[
+//MethodDeclaration
+   [not((@Name='main' and @Static=true())or ModifierList/Annotation/@SimpleName='PostConstruct'
+      or .//IfStatement//InfixExpression
+         [@Operator='=='][VariableAccess[pmd-java:typeIs('java.security.Provider')] and NullLiteral]
+   )]
+//ConstructorCall[pmd-java:typeIs('java.security.Provider') or pmd-java:typeIs('org.bouncycastle.jce.provider.BouncyCastleProvider')]
+                ]]></value>
+            </property>
+        </properties>
+        <example>
+            <![CDATA[
+import java.security.*;
+import javax.crypto.*;
+import org.bouncycastle.jce.provider.*;
+
+class Foo {
+    public Cipher initBlowfishBad() throws GeneralSecurityException {
+        Security.addProvider(new BouncyCastleProvider()); // bad
+        // build a Cipher
+    }
+
+    public Cipher initBlowfishGood() throws GeneralSecurityException {
+        Provider bouncyCastleProvider = Security.getProvider(BouncyCastleProvider.PROVIDER_NAME);
```

Comment:
Interesting question, in my recollection Security.getProvider is thread-safe somehow. It must be since it is a singleton and can be used heavily.
Variables are local, so safe.